### PR TITLE
New beautiful table

### PR DESCRIPTION
**Pull request by GeekaN2**

### CODE REVIEW

File: src/table.js [hunk 1]
```diff
@@ -1,74 +1,200 @@
...
-    const row = this._table.insertRow(index);
+    this.numberOfRows++;
+    let newRow;
+    let rowElem = create('div', [ CSS.row ]);
+
+    if (index > 0) {
+      let row = this.getRow(index);
 
-    this._fillRow(row);
+      if (row) {
```
Comment (neSpecc) [inline src/table.js:137]: comments needed

File: src/table.js [hunk 2]
```diff
@@ -1,74 +1,200 @@
...
+      const cell = this.getCell(i, index);
+
+      if (!cell) {
+        return;
+      }
+
+      cell.remove();
+    }
+
+    this.numberOfColumns--;
```
Comment (neSpecc) [inline src/table.js:169]: maybe its better to create a getter that will always return an actual value?

```
get columnsCount(){
  return ... // calculate from DOM
}
```

in this case you won't need to increment/decrement a property

Comment (neSpecc) [reply src/table.js:169]: or it can occur performance issue? Will it called in a loop?

File: src/table.js [hunk 3]
```diff
@@ -1,74 +1,200 @@
...
+    this.element.querySelector(`.${CSS.addColumn}`).innerHTML = svgPlusButton;
+    this.element.querySelector(`.${CSS.addRow}`).innerHTML = svgPlusButton;
+  }
+
   /**
    * get html element of table
    *
    * @returns {HTMLElement}
    */
   get htmlElement() {
```
Comment (neSpecc) [inline src/table.js:196]: name is not clear. What an element?

File: src/table.js [hunk 4]
```diff
@@ -1,74 +1,200 @@
...
-    return this._element;
+    return this.element;
   }
 
   /**
```
Comment (neSpecc) [inline src/table.js:200]: update doc

File: src/table.js [hunk 5]
```diff
@@ -77,148 +203,430 @@ export class Table {
    * @returns {HTMLElement}
    */
   get body() {
-    return this._table;
+    return this.table;
   }
 
   /**
-   * returns selected/editable cell
+   * Create wrapper with an additional interface
```
Comment (neSpecc) [inline src/table.js:210]: what does it mean?

File: src/table.js [hunk 6]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
-   * @returns {HTMLElement} tbody - where rows will be
+   * Add cells to a row
+   *
+   * @param {HTMLElement} row
    */
-  _createTableWrapper() {
-    return create('div', [ CSS.wrapper ], null, [ create('table', [ CSS.table ]) ]);
+  fillRow(row) {
+    for (let i = 1; i <= this.numberOfColumns; i++) {
+      const newCell = create('div', [ CSS.cell ], { contenteditable: !this.readOnly });
```
Comment (neSpecc) [inline src/table.js:231]: its the second place where you're creating a cell. Can you do it in a single place?

File: src/table.js [hunk 7]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+  hangEvents() {
+    // Update toolboxes position depending on the mouse movements
+    this.table.addEventListener('mousemove', (event) => {
+      const { row, column } = this.hoveredCell(event);
+
+      this.updateToolboxesPosition(row, column);
+
+      // set the listener to close toolboxes when click outside
+      document.addEventListener('click', this.clickOutsideWrapperListener);
+    }, true);
```
Comment (neSpecc) [inline src/table.js:249]: passive: true

File: src/table.js [hunk 8]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
-    return create('div', [ CSS.inputField ], { contenteditable: !this.readOnly });
+  hangEvents() {
+    // Update toolboxes position depending on the mouse movements
+    this.table.addEventListener('mousemove', (event) => {
+      const { row, column } = this.hoveredCell(event);
+
+      this.updateToolboxesPosition(row, column);
+
+      // set the listener to close toolboxes when click outside
+      document.addEventListener('click', this.clickOutsideWrapperListener);
```
Comment (neSpecc) [inline src/table.js:248]: you're adding a new click event on every mouse move? really?

File: src/table.js [hunk 9]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+    }, true);
+
+    // Add a column to the end
+    this.element.querySelector(`.${CSS.addColumn}`).addEventListener('click', (event) => {
```
Comment (neSpecc) [inline src/table.js:252]: don't use query selector in a methods. Save this button to the variable

File: src/table.js [hunk 10]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+      this.addColumn();
+      this.hideEverything();
```
Comment (neSpecc) [inline src/table.js:254]: name is not clear

File: src/table.js [hunk 11]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+    });
+
+    // Add a row to the bottom
+    this.element.querySelector(`.${CSS.addRow}`).addEventListener('click', (event) => {
```
Comment (neSpecc) [inline src/table.js:258]: don't add such a many event listeners. Use delegation. Add a single listener to the document and detect clicked element in there.


```js
const addRowClicked = event.target.closest(`.{CSS.addRow}`);

if (addRowClicked){
  this.addRow()
}
```

File: src/table.js [hunk 12]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+      this.addRow();
+      this.hideEverything();
+    });
+
+    // Add a column to the right
+    this.toolbox.toolboxColumn.querySelector(`.${CSS.toolboxAddColumnRight}`).addEventListener('click', event => {
+      event.stopPropagation();
+
+      this.addColumn(this.hoveredColumn + 1);
```
Comment (neSpecc) [inline src/table.js:267]: it is not clear , where this.hoveredColumn is updated?

File: src/table.js [hunk 13]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+      this.hideAndUnselect();
+    });
+
+    // Add a column to the left
```
Comment (neSpecc) [inline src/table.js:271]: use block comments

File: src/table.js [hunk 14]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+    this.toolbox.toolboxColumn.querySelector(`.${CSS.toolboxDeleteColumn}`).addEventListener('click', event => {
+      event.stopPropagation();
+
+      if (this.showDeleteColumnConfirmation) {
+        this.deleteColumn(this.hoveredColumn);
+        this.hideEverything();
+        this.showDeleteColumnConfirmation = false;
+      } else {
+        this.toolbox.setDeleteColumnConfirmation();
+        this.showDeleteColumnConfirmation = true;
```
Comment (neSpecc) [inline src/table.js:305]: move it to the `this.toolbox.setDeleteColumnConfirmation();`

it is related to the Toolbox class.

File: src/table.js [hunk 15]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+        this.showDeleteRowConfirmation = false;
+      } else {
+        this.toolbox.setDeleteRowConfirmation();
+        this.showDeleteRowConfirmation = true;
+      }
+    });
+
+    // Open/close toolbox row menu
+    this.toolbox.toolboxRow.addEventListener('click', event => {
+      this.unselectColumn();
```
Comment (neSpecc) [inline src/table.js:325]: move handlers to the separate methods. describe their workflow

File: src/table.js [hunk 16]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+      this.showDeleteColumnConfirmation = false;
+
+      this.selectColumn(this.hoveredColumn);
+      this.toolbox.openToolboxColumnMenu();
+      this.element.addEventListener('click', this.clickOutsideMenuListener, true);
+    });
+
+    // Controls some buttons inside the table
+    this.table.onkeypress = (event) => {
+      if (event.key == 'Enter' && event.shiftKey) {
```
Comment (neSpecc) [inline src/table.js:367]: add comment

File: src/table.js [hunk 17]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+        return true;
+      }
+
+      if (event.key == 'Enter') {
+        if (this.focusedCell.row != this.numberOfRows) {
```
Comment (neSpecc) [inline src/table.js:372]: move to the method

File: src/table.js [hunk 18]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+          this.focusCell(this.focusedCell);
+        } else {
+          this.addRow();
+          this.focusedCell.row += 1;
+          this.focusCell(this.focusedCell);
+          this.updateToolboxesPosition(0, 0);
+        }
+      }
+
+      return event.key != 'Enter';
```
Comment (neSpecc) [inline src/table.js:383]: add a comment

File: src/table.js [hunk 19]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
+    };
+
+    // Controls some buttons inside the table
+    this.table.addEventListener('keydown', (event) => {
+      if (event.key == 'Tab') {
```
Comment (neSpecc) [inline src/table.js:388]: add a comment

File: src/table.js [hunk 20]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
-   * @private
-   * @param {FocusEvent} event
+   * Makes the first row headings
+   *
+   * @param {boolean} withHeadings - use headings row or not
    */
-  _focusEditField(event) {
-    if (!event.target.classList.contains(CSS.inputField)) {
-      return;
+  useHeadings(withHeadings) {
```
Comment (neSpecc) [inline src/table.js:519]: name is not clear

File: src/table.js [hunk 21]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
-   * @param {FocusEvent} event
+   * Add effect of a selected row
+   *
+   * @param {number} index
    */
-  _blurEditField(event) {
-    if (!event.target.classList.contains(CSS.inputField)) {
-      return;
+  selectRow(index) {
+    this.lastSelectedRow = index;
```
Comment (neSpecc) [inline src/table.js:533]: you have an `if` statement, so move this setter there

File: src/table.js [hunk 22]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
-   * @private
-   * @returns {HTMLElement} - the area
+   * Hangs the necessary handlers to events
    */
-  _createContenteditableArea() {
-    return create('div', [ CSS.inputField ], { contenteditable: !this.readOnly });
+  hangEvents() {
+    // Update toolboxes position depending on the mouse movements
+    this.table.addEventListener('mousemove', (event) => {
+      const { row, column } = this.hoveredCell(event);
```
Comment (neSpecc) [inline src/table.js:243]: this method has many calculations, so please optimise it. Add a throttling, make listener a passive and so on.

File: src/table.js [hunk 23]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
   /**
-   * @private
-   * @returns {HTMLElement} - the area
+   * Hangs the necessary handlers to events
    */
-  _createContenteditableArea() {
-    return create('div', [ CSS.inputField ], { contenteditable: !this.readOnly });
+  hangEvents() {
+    // Update toolboxes position depending on the mouse movements
+    this.table.addEventListener('mousemove', (event) => {
```
Comment (neSpecc) [inline src/table.js:242]: be aware of using such events, because they're fired many times in a row. Optimise them.

File: src/table.js [hunk 24]
```diff
@@ -77,148 +203,430 @@ export class Table {
...
-   * @param {MouseEvent} event
+   * Calculates the row and column that the cursor is currently hovering over
+   *
+   * @param {Event} event - mousemove event
+   * @returns hovered cell coordinates as an integer row and column
    */
-  _mouseEnterInDetectArea(event) {
-    if (!event.target.classList.contains(CSS.area)) {
-      return;
+  hoveredCell(event) {
```
Comment (neSpecc) [inline src/table.js:598]: please optimise this method

File: src/toolbox.js [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+   * Creating a tooolbox row menu
+   *
+   * @returns {HTMLElement} - row menu
+   */
+  createRowMenu() {
+    let addRowAboveText = create('span');
+    let addRowBelowText = create('span');
+    let deleteRowText = create('span');
+
+    addRowAboveText.textContent = 'Add row above';
```
Comment (neSpecc) [inline src/toolbox.js:86]: use i18n

File: src/toolbox.js [hunk 26]
```diff
@@ -0,0 +1,291 @@
...
+
+  /**
+   * Remove the class to confirm deletion for the row menu
+   */
+  unsetDeleteRowConfirmation() {
+    this.toolboxRowMenu.querySelector(`.${CSS.toolboxDelete}`).classList.remove(CSS.deleteConfirm);
+  }
+
+  /**
+   * Change column toolbox position
```
Comment (neSpecc) [inline src/toolbox.js:243]: what does it mean?

File: src/toolbox.js [hunk 27]
```diff
@@ -0,0 +1,291 @@
...
+
+    if (numberOfColumns == 1) {
+      this.hideDeleteColumnButton();
+    } else {
+      this.unhideDeleteColumnButton();
+    }
+  }
+
+  /**
+   * Change row toolbox position
```
Comment (neSpecc) [inline src/toolbox.js:265]: what does it mean "row position"?

**Review by neSpecc — CHANGES_REQUESTED** [review]

**Review by neSpecc — CHANGES_REQUESTED** [review]
where is the `destroy()` method?

File: src/table.js [hunk 28]
```diff
@@ -77,148 +221,534 @@ export class Table {
...
+      }
+
+      if (addColumnClicked) {
+        this.addColumn();
+        this.hideEverything();
+      }
+    });
+
+    // Controls toolbox with adding and deleting columns
+    this.toolbox.toolboxColumn.addEventListener('click', (event) => {
```
Comment (neSpecc) [inline src/table.js:296]: move handler

File: src/table.js [hunk 29]
```diff
@@ -77,148 +221,534 @@ export class Table {
...
+      const addColumnRightClicked = event.target.closest(`.${CSS.toolboxAddColumnRight}`);
+      const addColumnLeftClicked = event.target.closest(`.${CSS.toolboxAddColumnLeft}`);
+      const deleteColumnClicked = event.target.closest(`.${CSS.toolboxDeleteColumn}`);
+
+      if (addColumnRightClicked) {
+        this.addColumn(this.hoveredColumn + 1);
+        this.hideAndUnselect();
+      }
+
+      if (addColumnLeftClicked) {
```
Comment (neSpecc) [inline src/table.js:309]: `if else` or `return`

File: src/table.js [hunk 30]
```diff
@@ -77,148 +221,534 @@ export class Table {
...
+        }
+      }
+
+      // Open/close toolbox column menu
+      if (toolboxRowIconClicked) {
+        this.unselectColumnAndHideMenu();
+
+        if (this.hoveredRow == this.lastSelectedRow) {
+          this.unselectRowAndHideMenu();
+
```
Comment (neSpecc) [inline src/table.js:376]: can we move these lines to some util? They are similar with previous

File: src/table.js [hunk 31]
```diff
@@ -77,148 +221,534 @@ export class Table {
...
+      }
+    });
+
+    // Controls some of the keyboard buttons inside the table
+    this.table.onkeypress = (event) => {
+      if (event.key == 'Enter' && event.shiftKey) {
+        return true;
+      }
+
+      if (event.key == 'Enter') {
```
Comment (neSpecc) [inline src/table.js:391]: duplicated statement

File: src/table.js [hunk 32]
```diff
@@ -77,148 +221,534 @@ export class Table {
...
+          this.focusCell(this.focusedCell);
+          this.updateToolboxesPosition(0, 0);
+        }
+      }
+
+      return event.key != 'Enter';
+    };
+
+    // Controls some of the keyboard buttons inside the table
+    this.table.addEventListener('keydown', (event) => {
```
Comment (neSpecc) [inline src/table.js:407]: why `keydown` and `keypress` are different handlers?

File: src/table.js [hunk 33]
```diff
@@ -77,148 +221,534 @@ export class Table {
...
+      if (event.key == 'Tab') {
+        event.stopPropagation();
+      }
+    });
+
+    // Determine the position of the cell in focus
+    this.table.addEventListener('focusin', event => {
+      const cell = event.target;
+      const row = cell.parentElement;
```
Comment (neSpecc) [inline src/table.js:416]: use method `getRowByCell()`

File: src/table.js [hunk 34]
```diff
@@ -77,148 +221,534 @@ export class Table {
...
+
+          break;
+        }
+
+        totalIterations++;
+      }
     }
+
+    // Looking for hovered row using binsearch
+    if (y >= 0 && !hoveredRow) {
```
Comment (neSpecc) [inline src/table.js:651]: do not repeat yourself

File: src/table.js [hunk 35]
```diff
@@ -77,148 +221,534 @@ export class Table {
...
+    const { width, height, x, y } = getRelativeCoords(this.table, event);
+
+    // Looking for hovered column using binsearch
+    if (x >= 0 && !hoveredColumn) {
+      let totalIterations = 0;
+
+      while (leftBorder < rightBorder && totalIterations < 10) {
+        const mid = Math.ceil((leftBorder + rightBorder) / 2);
+        const cell = this.getCell(1, mid);
+        const { fromRightBorder, fromLeftBorder } = getRelativeCoordsOfTwoElems(this.table, cell);
```
Comment (neSpecc) [inline src/table.js:634]: 1. describe why you need this? 
2. Can you calculate and cache values before iteration?

File: src/toolbox.js [hunk 36]
```diff
@@ -0,0 +1,291 @@
...
+  hidden: 'tc-hidden',
+  displayNone: 'tc-display-none',
+  toolboxRow: 'tc-toolbox-row',
+  toolboxRowMenu: 'tc-toolbox-row__menu',
+  toolboxColumn: 'tc-toolbox-column',
+  toolboxColumnMenu: 'tc-toolbox-column__menu',
+  toolboxAddColumnRight: 'tc-toolbox-add-column-right',
+  toolboxAddColumnLeft: 'tc-toolbox-add-column-left',
+  toolboxAddRowAbove: 'tc-toolbox-add-row-above',
+  toolboxAddRowBelow: 'tc-toolbox-add-row-below',
```
Comment (neSpecc) [inline src/toolbox.js:19]: do not copy indentifiers

File: src/toolbox.js [hunk 37]
```diff
@@ -0,0 +1,291 @@
...
+
+/**
+ * @constructor
+ */
+export class Toolbox {
+  /**
+   * Creates toolbox buttons and toolbox menus
+   */
+  constructor() {
+    this.toolboxRow = this.createToolboxRow();
```
Comment (neSpecc) [inline src/toolbox.js:36]: separate to different classes

File: src/toolbox.js [hunk 38]
```diff
@@ -0,0 +1,291 @@
...
+  createRowMenu() {
+    let addRowAboveText = create('span');
+    let addRowBelowText = create('span');
+    let deleteRowText = create('span');
+
+    addRowAboveText.textContent = 'Add row above';
+    addRowBelowText.textContent = 'Add row below';
+    deleteRowText.textContent = 'Delete row';
+
+    let addRowAbove = create('div', [CSS.toolboxAddRowAbove, CSS.toolboxOption]);
```
Comment (neSpecc) [inline src/toolbox.js:90]: use cycle for rendering

File: src/toolbox.js [hunk 39]
```diff
@@ -0,0 +1,291 @@
...
+    ]);
+
+    return toolboxColumnMenu;
+  }
+
+  /**
+   * Hide delete column button for event when we only have one column left
+   */
+  hideDeleteColumnButton() {
+    this.toolboxColumnMenu.querySelector(`.${CSS.toolboxDeleteColumn}`).classList.add(CSS.displayNone);
```
Comment (neSpecc) [inline src/toolbox.js:156]: do not use DOM search

File: src/toolbox.js [hunk 40]
```diff
@@ -0,0 +1,291 @@
...
+   * @param {number} numberOfColumns - number of columns in the table
+   * @param {number} column - current column, if 0 then hide toolbox
+   */
+  updateToolboxColumnPosition(numberOfColumns = 0, column = this.column) {
+    this.column = column;
+
+    if (this.column <= 0 || this.column > numberOfColumns) {
+      this.toolboxColumn.style.opacity = '0';
+    } else {
+      this.toolboxColumn.style.cssText = 'opacity: 1; ' + `left: calc((100% - 35px) / (${numberOfColumns} * 2) * (1 + (${column} - 1) * 2))`;
```
Comment (neSpecc) [inline src/toolbox.js:254]: comment required

**Review by neSpecc — CHANGES_REQUESTED** [review]

**Review by neSpecc — APPROVED** [review]

**Review by talyguryn — APPROVED** [review]

**Pull request merged**